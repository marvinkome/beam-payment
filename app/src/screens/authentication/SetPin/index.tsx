import React, { useContext, useState } from "react"
import { useNavigation } from "@react-navigation/native"
import { gql, useMutation } from "@apollo/client"
import { SetPinScreen } from "./SetPin"
import { ToastAndroid } from "react-native"
import { AuthContext } from "libs/auth-context"

export const SAVE_PIN = gql`
    mutation SetPin($pin: String!) {
        setPin(pin: $pin) {
            success
            responseMessage
            user {
                id
            }
        }
    }
`

function useSavePin() {
    const authContext = useContext(AuthContext)
    const [savePin] = useMutation(SAVE_PIN)
    const [savingPin, setSavingPin] = useState(false)

    return {
        savingPin,
        savePin: async (pin: string) => {
            setSavingPin(true)
            try {
                const savePinResp = await savePin({ variables: { pin } })
                const { success, responseMessage } = savePinResp?.data?.setPin

                if (!success) {
                    return ToastAndroid.show(responseMessage, ToastAndroid.SHORT)
                }

                authContext?.signIn()
            } catch (e) {
                console.log(e)
                ToastAndroid.show("Error saving pin. Try again", ToastAndroid.SHORT)
            }
        },
    }
}

export function SetPin() {
    const [pin, setPin] = useState("")
    const { savePin, savingPin } = useSavePin()

    return (
        <SetPinScreen
            loading={savingPin}
            pin={pin}
            setPin={setPin}
            onContinue={() => savePin(pin)}
        />
    )
}
