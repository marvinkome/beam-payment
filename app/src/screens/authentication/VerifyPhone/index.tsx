import React, { useState } from "react"
import auth from "@react-native-firebase/auth"
import { ToastAndroid } from "react-native"
import { smsConfirmationObj } from "store/authStore"
import { useAuthentication } from "hooks/login"
import { VerifyPhoneScreen } from "./VerifyPhone"

export function VerifyPhone() {
    const [code, setCode] = useState("")
    const [verifingCode, setVerifingCode] = useState(false)
    const { signIn } = useAuthentication()
    const { confirmation, phoneNumber } = smsConfirmationObj()

    const onVerify = async () => {
        setVerifingCode(true)

        try {
            await confirmation?.confirm(code)
            const idToken = await auth().currentUser?.getIdToken()

            if (!idToken) {
                throw new Error()
            }

            await signIn(idToken)
            setVerifingCode(false)
        } catch (e) {
            setVerifingCode(false)
            ToastAndroid.show("Invalid code", ToastAndroid.SHORT)
        }
    }

    return (
        <VerifyPhoneScreen
            phoneNumber={phoneNumber}
            onVerify={onVerify}
            code={code}
            setCode={setCode}
            loading={verifingCode}
        />
    )
}
