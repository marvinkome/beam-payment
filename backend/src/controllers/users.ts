import { IUser } from "models/users"

export async function setPin(data: { pin: string }, user: IUser | null) {
    // set pin and return user
    if (!user) {
        return
    }

    user.pin = data.pin

    try {
        await user.save()
        return {
            success: true,
            user,
        }
    } catch (e) {
        return {
            success: false,
            responseMessage: "Error saving pin. Please try again",
        }
    }
}
