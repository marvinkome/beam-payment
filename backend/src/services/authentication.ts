import User from "models/users"

export async function createUserAccount(phoneNumber: string, firebaseId?: string) {
    const user = new User()

    user.phoneNumber = phoneNumber
    user.firebaseId = firebaseId

    return user.save()
}
