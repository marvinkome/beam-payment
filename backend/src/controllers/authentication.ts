import { auth } from "firebase-admin"
import Logger from "loaders/logger"
import { findOrCreateUserAccount } from "services/authentication"

export async function authenticateUser(data: { idToken: string }) {
    try {
        const { phone_number, uid } = await auth().verifyIdToken(data.idToken)

        // create user and return jwt token
        const { user, token } = await findOrCreateUserAccount(phone_number!, uid)

        return {
            success: true,
            user,
            token,
        }
    } catch (e) {
        // TODO:: use sentry
        Logger.error(e)
        return {
            success: false,
            responseMessage: "Error creating account",
        }
    }
}
