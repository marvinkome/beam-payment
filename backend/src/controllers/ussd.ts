import express from "express"
import { findOrCreateUserAccount } from "services/authentication"
import { UssdService } from "services/ussd"

export async function handleUSSDCode(req: express.Request, res: express.Response) {
    const { phoneNumber, text } = req.body

    const { user } = await findOrCreateUserAccount(phoneNumber)
    const ussdService = new UssdService(user)

    res.set("Content-Type", "text/plain")
    return res.send(await ussdService.handleUSSDCode(text))
}
