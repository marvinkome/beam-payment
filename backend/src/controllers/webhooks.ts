import * as Sentry from "@sentry/node"
import config from "config"
import express from "express"
import Logger from "loaders/logger"
import Transaction from "models/transactions"
import { IUser } from "models/users"
import { UserService } from "services/user"

export async function onFailedWithdrawal(req: express.Request, res: express.Response) {
    // Logs
    // Logger.info("Transfer webhook received", req.body)

    const hash = req.headers["verif-hash"]
    if (!hash) return res.sendStatus(200)

    const secret_hash = config.flutterwaveSecretHash
    if (hash !== secret_hash) res.sendStatus(200)

    // if transfer wasn't successful, return the original amount
    const response = req.body
    Sentry.addBreadcrumb(response)

    if (
        response?.data?.status === "SUCCESSFUL" &&
        response?.data?.complete_message === "Successful"
    ) {
        return res.sendStatus(200)
    }

    const transactionRef = await Transaction.findOne({ transactionId: response.data?.reference })

    if (!transactionRef) {
        return res.sendStatus(200)
    }

    // refund user
    const { from } = await transactionRef.populate("from").execPopulate()
    const userService = new UserService(from! as IUser)
    await userService.revertTransaction(transactionRef)

    res.sendStatus(200)
}
