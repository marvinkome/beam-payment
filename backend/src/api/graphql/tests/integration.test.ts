import { gql } from "apollo-server-express"
import { createTestClient } from "apollo-server-testing"
import { authenticateUser } from "controllers/authentication"
import { constructTestServer } from "./__utils"

jest.mock("controllers/authentication")

describe("Mutation", () => {
    describe("authenticateUser", () => {
        test("success: true state", async () => {
            // @ts-ignore
            authenticateUser.mockImplementationOnce(() => ({
                success: true,
                token: "token",
                user: {
                    id: "userId",
                    phoneNumber: "phoneNumber",
                },
            }))

            const server = constructTestServer()
            const { mutate } = createTestClient(server)

            const response = await mutate({
                mutation: gql`
                    mutation AuthenticateUser($idToken: String!) {
                        authenticateUser(idToken: $idToken) {
                            success
                            token
                            user {
                                id
                            }
                        }
                    }
                `,
                variables: { idToken: "id-token" },
            })

            expect(response.errors).toBeUndefined()
            expect(response.data?.authenticateUser.token).toBe("token")
            expect(response).toMatchSnapshot()
        })

        test("success: false state", async () => {
            // @ts-ignore
            authenticateUser.mockImplementationOnce(() => ({
                success: false,
                responseMessage: "Error creating user",
            }))

            const server = constructTestServer()
            const { mutate } = createTestClient(server)

            const response = await mutate({
                mutation: gql`
                    mutation AuthenticateUser($idToken: String!) {
                        authenticateUser(idToken: $idToken) {
                            success
                            responseMessage
                            token
                            user {
                                id
                            }
                        }
                    }
                `,
                variables: { idToken: "id-token" },
            })

            expect(response.errors).toBeUndefined()
            expect(response.data?.authenticateUser.responseMessage).toBe("Error creating user")
            expect(response.data?.authenticateUser.token).toBe(null)
            expect(response).toMatchSnapshot()
        })
    })
})
