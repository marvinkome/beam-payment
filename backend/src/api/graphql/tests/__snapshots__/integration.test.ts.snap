// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation authenticateUser success: false state 1`] = `
Object {
  "data": Object {
    "authenticateUser": Object {
      "isNewAccount": null,
      "responseMessage": "Error creating user",
      "success": false,
      "token": null,
      "user": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation authenticateUser success: true state 1`] = `
Object {
  "data": Object {
    "authenticateUser": Object {
      "isNewAccount": true,
      "responseMessage": null,
      "success": true,
      "token": "token",
      "user": Object {
        "id": "userId",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation setPin success 1`] = `
Object {
  "data": Object {
    "setPin": Object {
      "responseMessage": null,
      "success": true,
      "user": Object {
        "id": "0",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
